lib/srv/alpnproxy/kube.go

```diff
@@ -66,13 +99,92 @@ func (m *KubeMiddleware) OverwriteClientCerts(req *http.Request) ([]tls.Certific
 		return nil, trace.BadParameter("expect a TLS request")
 	}
 
+	m.certsMu.RLock()
 	cert, ok := m.certs[req.TLS.ServerName]
+	m.certsMu.RUnlock()
 	if !ok {
 		return nil, trace.NotFound("no client cert found for %v", req.TLS.ServerName)
 	}
+
+	var err error
+	cert, err = m.reissueCertIfExpired(req.Context(), cert, req.TLS.ServerName)
+	if err != nil {
+		// If we exceeded timeout, we return an error that will try to get user attention to the local proxy
+		if errors.Is(err, ErrUserInputRequired) {
+			return nil, trace.Wrap(&apierrors.StatusError{
+				ErrStatus: metav1.Status{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "Status",
+						APIVersion: "v1",
+					},
+					Status:  metav1.StatusFailure,
+					Code:    http.StatusGatewayTimeout,
+					Reason:  metav1.StatusReasonTimeout,
+					Message: "Local Kube proxy requires user input to continue",
+				},
+			})
+		}
+		return nil, trace.Wrap(err)
+	}
+
 	return []tls.Certificate{cert}, nil
 }
 
+// ErrUserInputRequired returned when user's input required to relogin and/or reissue new certificate.
+var ErrUserInputRequired = errors.New("user input required")
+
+// reissueCertIfExpired checks if provided certificate has expired and reissues it if needed, otherwise it returns original certificate.
+func (m *KubeMiddleware) reissueCertIfExpired(ctx context.Context, cert tls.Certificate, serverName string) (tls.Certificate, error) {
+	x509Cert, err := utils.TLSCertLeaf(cert)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
```

Comment:
I don't think `issueKubeCert` returns a cert with a leaf set. maybe we can set leaf there to avoid parsing every time.

Reply:
Done 👍